db/python/tables/meta_table.py

```diff
@@ -0,0 +1,144 @@
+        parquet file.
+        """
+
+        rows = await self.connection.fetch_all(
+            query,
+            {
+                'project': project,
+                'primary_external_org': PRIMARY_EXTERNAL_ORG,
+                'primary_external_org_sentinel': EXTERNAL_ORG_SENTINEL,
+            },
+        )
+
+        if len(rows) == 0:
+            return None
+
+        duck = duckdb.connect()
+        meta_rows_str = StringIO()
+        external_id_rows_str = StringIO()
+        main_rows: list[dict[str, Any]] = []
+
+        for row in rows:
+            if has_meta:
+                meta_rows_str.write((row['meta'] or 'null') + '\n')
+
+            if has_external_ids:
+                external_id_rows_str.write(
+                    (row['external_ids'] or f'{{"{EXTERNAL_ORG_SENTINEL}": null}}')
+                    + '\n'
+                )
+
+            main_rows.append(row_getter(row))
+
+        duck.register('main_table', pa.Table.from_pylist(main_rows))
+
+        # Seek str buffers back to beginning
+        meta_rows_str.seek(0)
+        external_id_rows_str.seek(0)
+
+        # parse the json for the meta table
+        meta_columns = ''
+        meta_join = ''
+        if has_meta:
+            duck.register(
+                'meta_table',
+                duck.read_json(
+                    meta_rows_str,
+                    map_inference_threshold=-1,
+                    format='newline_delimited',
+                ).arrow(),
+            )
+            # Prefix all meta columns with `meta_` to avoid clashes with the main table
+            meta_columns = """
+                ,COLUMNS(mt.*) as 'meta_\\0'
+            """
+            meta_join = 'POSITIONAL JOIN meta_table mt'
+
+        # parse the json for the external_id table
+        external_id_columns = ''
+        external_id_join = ''
+        if has_external_ids:
+            duck.register(
+                'external_id_table',
+                duck.read_json(
+                    external_id_rows_str,
+                    format='newline_delimited',
+                ).arrow(),
+            )
+
+            # call the primary external_id column `external_id` and prefix all other
+            # columns with `external_id_` to avoid clashes with the main table
+            external_id_columns = f"""
+                ,et.{EXTERNAL_ORG_SENTINEL} as external_id
+                ,COLUMNS(et.* EXCLUDE {EXTERNAL_ORG_SENTINEL}) as 'external_id_\\0'
+            """
+            external_id_join = 'POSITIONAL JOIN external_id_table et'
+
+        # combine the three tables together, if they are populated
+        combined = duck.query(f"""
+            SELECT
+                m.*
+                {external_id_columns}
+                {meta_columns}
+
+            FROM main_table m
+            {external_id_join}
+            {meta_join}
+        """)
+
+        buffer = BytesIO()
+        pq.write_table(combined.arrow(), buffer)
+
+        duck.close()
```

Comment:
Just a query. The use of duckdb is a personal preference? Or does it have particular functionality to enable the creation of the parquet file? Or is the pyarrow.parquet module here the one actually doing the write to the parquet file and would take a table in a sensible format?

Reply:
The pyarrow module is doing the actual writing of the parquet file. Duckdb can write parquet files but only to disk and not into a filelike BytesIO object. Because duckdb and pyarrow both operate on arrow tables the data isn't copied in memory and it is really fast to interoperate between the two tools. Arrow is a really cool specification and has implementations in a bunch of languages. It's becoming a bit of a standard for working with in-memory tabular data.

In theory this could all be done just with pyarrow, but I'm mostly using duckdb here for its json parsing abilities. The JSON we have in metamist isn't very well structured, and lots of tools like Polars will try and infer the schema of the first chunk of rows. Because our meta blobs are a bit all over the place it's very possible that row 10,000 will have some columns that no other rows have. Duckdb is really fast at parsing nested and complex json.

Reply:
Good explanation. Yeah I was just wondering the decision behind these choices more than anything. I've seen some use of pyarrow before. Duckdb is new but very cool.